boltdb/adaptor_test.go

```diff
@@ -0,0 +1,204 @@
+
+			Expect(ok).Should(BeTrue())
+			Expect(err).ShouldNot(HaveOccurred())
+
+			v, err = adaptor.ResourceVersion(
+				context.Background(),
+				[]byte("<resource>"),
+			)
+			Expect(err).ShouldNot(HaveOccurred())
+			Expect(v).To(Equal([]byte("<version 02>")))
+
+			By("discarding a resource if the next resource version is empty")
+			ok, err = adaptor.HandleEvent(
+				context.Background(),
+				[]byte("<resource>"),
+				[]byte("<version 02>"),
+				nil,
+				nil,
+				fixtures.MessageA3,
+			)
+			Expect(ok).Should(BeTrue())
+			Expect(err).ShouldNot(HaveOccurred())
+
+			v, err = adaptor.ResourceVersion(
+				context.Background(),
+				[]byte("<resource>"),
+			)
+			Expect(err).ShouldNot(HaveOccurred())
+			Expect(v).To(BeEmpty())
+		})
+
+		It("returns an error if MessageHandler's HandleEvent method fails", func() {
+			terr := errors.New("handle event test error")
+
+			handler.HandleEventCall = func(
+				context.Context,
+				*bolt.Tx,
+				dogma.ProjectionEventScope,
+				dogma.Message,
+			) error {
+				return terr
+			}
+
+			ok, err := adaptor.HandleEvent(
+				context.Background(),
+				[]byte("<resource>"),
+				nil,
+				[]byte("<version 01>"),
+				nil,
+				fixtures.MessageA1,
+			)
+
+			Expect(ok).Should(BeFalse())
+			Expect(err).Should(HaveOccurred())
+		})
+
+		It("returns false if the correct resource version in the database is incorrect", func() {
+			ok, err := adaptor.HandleEvent(
+				context.Background(),
+				[]byte("<resource>"),
+				nil,
+				[]byte("<version 01>"),
+				nil,
+				fixtures.MessageA1,
+			)
+
+			Expect(ok).Should(BeTrue())
+			Expect(err).ShouldNot(HaveOccurred())
+
+			ok, err = adaptor.HandleEvent(
+				context.Background(),
+				[]byte("<resource>"),
+				[]byte("<incorrect current version>"),
+				[]byte("<version 02>"),
+				nil,
+				fixtures.MessageA2,
+			)
+
+			Expect(ok).Should(BeFalse())
+			Expect(err).ShouldNot(HaveOccurred())
+		})
+	})
+
+	Context("func ResourceVersion()", func() {
+		It("returns a resource version", func() {
+
+			ok, err := adaptor.HandleEvent(
+				context.Background(),
+				[]byte("<resource>"),
+				nil,
+				[]byte("<version 01>"),
+				nil,
+				fixtures.MessageA1,
+			)
+
+			Expect(ok).Should(BeTrue())
+			Expect(err).ShouldNot(HaveOccurred())
+
+			v, err := adaptor.ResourceVersion(
+				context.Background(),
+				[]byte("<resource>"),
+			)
+			Expect(err).ShouldNot(HaveOccurred())
+			Expect(v).To(Equal([]byte("<version 01>")))
+		})
+	})
+
+	Context("func CloseResource()", func() {
+		It("removes a resource version", func() {
+			ok, err := adaptor.HandleEvent(
+				context.Background(),
+				[]byte("<resource>"),
+				nil,
+				[]byte("<version 01>"),
+				nil,
+				fixtures.MessageA2,
+			)
+
+			Expect(ok).Should(BeTrue())
+			Expect(err).ShouldNot(HaveOccurred())
+
+			err = adaptor.CloseResource(
+				context.Background(),
+				[]byte("<resource>"),
+			)
+			Expect(err).ShouldNot(HaveOccurred())
+
+			v, err := adaptor.ResourceVersion(
```

Comment:
Hey, is this what you were specifically referring to the other day when we were talking about testing using the public interface, vs the internal functions?

Reply:
That's pretty much it. I am using `adaptor.ResourceVersion() ` method to verify the correct entry in the database, whereas before the intent was to create a `database verifier` object that would independently verify entries in the database, without resorting to public interface methods.

Reply:
Ok yeah, I had just been trying to think of a way to test it to the letter of the spec but that's kind of impossible because it [states specifically](https://github.com/dogmatiq/dogma/blob/892967a8653cd330e0833f98ebdbd964db341e6c/projection.go#L103) the the behavior after closing a resource is undefined.

At any rate, in the context of this implementation we know that it should delete the resource entirely, so the test is fine as is.